### PR TITLE
[flink] support alter table for flink

### PR DESCRIPTION
**Pull request by zhangjun0x01**
*(Please specify the module before the PR name: [core] ... or [flink] ...)*

### Purpose

*Related to https://github.com/apache/incubator-paimon/issues/785*

### Tests

*org.apache.paimon.flink.SchemaChangeITCase*

### API and Format 

*(Does this change affect API or storage format)*

### Documentation

*(Does this change introduce a new feature)*

### CODE REVIEW

File: docs/content/how-to/altering-tables.md [hunk 1]
```diff
@@ -114,6 +114,15 @@ The following SQL adds two columns `c1` and `c2` to table `my_table`.
 
 {{< tabs "add-columns-example" >}}
 
+{{< tab "Flink" >}}
+
+```sql
+CREATE TABLE T (a STRING, b DOUBLE, c FLOAT);
+ALTER TABLE T ADD d INT;
+```
```
Comment (yuzelin) [inline docs/content/how-to/altering-tables.md:122]: Please not the words above: The following SQL adds two columns `c1` and `c2` to table `my_table`.
The example should be similar to spark's example.
Flink also support add multiple columns. See (from flink doc):
 ADD { <schema_component> | (<schema_component> [, ...]) }

File: docs/content/how-to/altering-tables.md [hunk 2]
```diff
@@ -150,6 +169,15 @@ The following SQL renames column `c0` in table `my_table` to `c1`.
 
 {{< tabs "rename-column-name-example" >}}
 
+{{< tab "Flink" >}}
+
+```sql
+ALTER TABLE T RENAME c TO c1;
```
Comment (yuzelin) [inline docs/content/how-to/altering-tables.md:175]: The following SQL renames column `c0` in table `my_table` to `c1.

File: docs/content/how-to/altering-tables.md [hunk 3]
```diff
@@ -165,6 +193,14 @@ The syntax is:
 
 {{< tabs "drop-columns-syntax" >}}
 
+{{< tab "Flink" >}}
+
+```sql
+ALTER TABLE T DROP e;
```
Comment (yuzelin) [inline docs/content/how-to/altering-tables.md:199]: Note that here is syntax, example is behind.

File: docs/content/how-to/altering-tables.md [hunk 4]
```diff
@@ -195,6 +231,16 @@ The following SQL sets column `coupon_info` to be nullable.
 
 {{< tabs "change-nullability-example" >}}
 
+{{< tab "Flink" >}}
+
+```sql
+CREATE TABLE T (a STRING PRIMARY KEY NOT ENFORCED, b FLOAT NOT NULL);
+ALTER TABLE T MODIFY b FLOAT;
+```
```
Comment (yuzelin) [inline docs/content/how-to/altering-tables.md:239]: 1. The following SQL sets column `coupon_info` to be nullable.
2. modify nullability, not type.

I think changing column type is missing, please also add a new part for `changing column type` for both spark and flink.

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java [hunk 5]
```diff
@@ -28,7 +32,211 @@
 /** ITCase for schema changes. */
 public class SchemaChangeITCase extends CatalogITCaseBase {
 
-    // TODO cover more cases once Flink supports more ALTER operations.
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java:31]: Please change to `TODO cover more cases` because some changes like `AddUniqueConstraint` are still not supported in this PR.

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java [hunk 6]
```diff
@@ -28,7 +32,211 @@
...
+        List<Row> result = sql("SHOW CREATE TABLE T");
+        assertThat(result.toString())
+                .contains(
+                        "CREATE TABLE `PAIMON`.`default`.`T` (\n"
+                                + "  `b` VARCHAR(2147483647),\n"
+                                + "  `a` VARCHAR(2147483647) NOT NULL,\n"
+                                + "  `c` VARCHAR(2147483647),\n"
+                                + "  `d` INT,\n"
+                                + "  `e` DOUBLE,");
+
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java:173]: Test: insert values before moving and read after moving

**Comment by SteNicholas** [comment]
@zhangjun0x01, could you rebase master branch to fix the failure of check licensing?

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java [hunk 7]
```diff
@@ -28,7 +32,211 @@
...
+                                + "  `b` VARCHAR(2147483647),\n"
+                                + "  `c` VARCHAR(2147483647),\n"
+                                + "  `d` INT,\n"
+                                + "  `e` DOUBLE,");
+
+        assertThatThrownBy(() -> sql("ALTER TABLE T MODIFY c DOUBLE"))
+                .getRootCause()
+                .isInstanceOf(IllegalStateException.class)
+                .hasMessage(
+                        "Column type c[STRING] cannot be converted to DOUBLE without loosing information.");
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java:156]: The message `Column type c[STRING] cannot be converted to DOUBLE without loosing information` is from
`SchemaManager` Line 291. I believe there is a typo error. I think it should be `without losing information`.

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java [hunk 8]
```diff
@@ -28,7 +32,211 @@
...
+        sql("ALTER TABLE T MODIFY e DOUBLE");
+        List<Row> result = sql("SHOW CREATE TABLE T");
+        assertThat(result.toString())
+                .contains(
+                        "CREATE TABLE `PAIMON`.`default`.`T` (\n"
+                                + "  `a` VARCHAR(2147483647) NOT NULL,\n"
+                                + "  `b` VARCHAR(2147483647),\n"
+                                + "  `c` VARCHAR(2147483647),\n"
+                                + "  `d` INT,\n"
+                                + "  `e` DOUBLE,");
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java:150]: Test: insert value before and read after.

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java [hunk 9]
```diff
@@ -28,7 +32,211 @@
...
+                                + ")");
+
+        sql("INSERT INTO T2 VALUES('paimon', 'bbb', 3.4, 6.7)");
+        result = sql("SELECT * FROM T2");
+        assertThat(result.toString())
+                .isEqualTo("[+I[null, aaa, 1.2, 3.4], +I[paimon, bbb, 3.4, 6.7]]");
+    }
+
+    @Test
+    public void testDropColumn() {
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/SchemaChangeITCase.java:91]: 1. Test: insert value before and read after.
2. Please add test for drop partition key and primary key (should throw exception). See `SparkSchemaEvolutionITCase`.

**Review by yuzelin — COMMENTED** [review]
Thanks for contribution~
I left some comments, please take a look.

File: docs/content/how-to/altering-tables.md (hunk 4 above)
Comment (zhangjun0x01) [reply docs/content/how-to/altering-tables.md:239]: > * The following SQL sets column `coupon_info` to be nullable.
> * modify nullability, not type.

I don't seem to have found any relevant examples to modify nullability in flink [doc](https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/alter/#modify)

Using an `alter table` syntax without `not null` to set a column to be nullable

Comment (yuzelin) [reply docs/content/how-to/altering-tables.md:239]: I asked flink community and the answer is 'alter table modify column_name data_type not null'. Can you have a try?

Comment (zhangjun0x01) [reply docs/content/how-to/altering-tables.md:239]: > I asked flink community and the answer is `alter table modify column_name data_type not null`. Can you have a try?

I have add a test in the `SchemaChangeITCase`, change the column to not null will throw an exception.

```
      assertThatThrownBy(() -> sql("ALTER TABLE T MODIFY c STRING NOT NULL"))
                .hasMessageContaining(
                        "Could not execute ALTER TABLE PAIMON.default.T\n"
                                + "  MODIFY `c` STRING NOT NULL");

```

**Comment by zhangjun0x01** [comment]
> @zhangjun0x01, could you rebase master branch to fix the failure of check licensing?

I update it

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 10]
```diff
@@ -222,52 +234,106 @@ public void createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ig
...
 
-        oldProperties
-                .keySet()
-                .forEach(
-                        k -> {
-                            if (!newTable.getOptions().containsKey(k)) {
-                                changes.add(SchemaChange.removeOption(k));
-                            }
-                        });
-
```
[the remark this answers is not in the document]
Comment (zhangjun0x01) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java]: > We cannot delete this because < 1.17 version still use the old style `alterTable` API. How about don't make `alterTable(ObjectPath tablePath, CatalogBaseTable newTable, boolean ignoreIfNotExists)` call the new `alterTable`? To verify that old `alter table set option` still works, we can copy old tests in `ShemaChangeITCase` to Flink 1.16 module.

I copy the previous `ShemaChangeITCase ` and `FlinkCatalog` to flink 1.14, 1.15, 1.16 , in  flink 1.17 ，I think it's best to unify the logic, use  `FlinkCatalog#toSchemaChange` to get the change, what do you think about ?

**Comment by zhangjun0x01** [comment]
I addressed all the comments, could you help me review it again ? @yuzelin  ,thanks

File: docs/content/how-to/altering-tables.md (hunk 4 above)
Comment (yuzelin) [reply docs/content/how-to/altering-tables.md:247]: Ok, seems flink only support change `not null` to `nullable`.
I have some thought about this. If the column is nullable before, we can insert null value. Then if we change it to not null, the previous null value is conflict to the `not null` definition. So I think flink's behavior is reasonable.

**Comment by yuzelin** [comment]
Hi, @zhangjun0x01 , thanks very much. I see that you have modified the test and documentation. The modification of test and docs look good to me.

But I am afraid of that I cannot agree to copy `FlinkCatalog` to 1.14-1.16 modules to unify the logic, because there is too many redundant code (now this PR has 2600+). I have test my thought based on this PR in this commit: https://github.com/apache/incubator-paimon/commit/ba1f31bf720ed2091598cf3363eb249ed9c14c78
It works good. I think that keep code base clean is important.

cc @JingsongLi

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (yuzelin) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java]: I left a comment at the bottom about this.

**Comment by yuzelin** [comment]
Hi, @zhangjun0x01, is there any progress?

**Comment by zhangjun0x01** [comment]
> Hi, @zhangjun0x01, is there any progress?

sorry for the late reply，I update it，thanks for your suggestion.

**Comment by zhangjun0x01** [comment]
image

**Comment by zhangjun0x01** [comment]
@JingsongLi @yuzelin  could you help me review it again ? thanks

File: paimon-flink/paimon-flink-1.16/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 11]
```diff
@@ -0,0 +1,564 @@
...
+
+import static org.apache.flink.table.descriptors.Schema.SCHEMA;
+import static org.apache.flink.table.factories.FactoryUtil.CONNECTOR;
+import static org.apache.flink.table.types.utils.TypeConversions.fromLogicalToDataType;
+import static org.apache.paimon.CoreOptions.PATH;
+import static org.apache.paimon.flink.LogicalTypeConversion.toDataType;
+import static org.apache.paimon.flink.LogicalTypeConversion.toLogicalType;
+
+/** Catalog for paimon. */
+public class FlinkCatalog extends AbstractCatalog {
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-1.16/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:71]: This is not necessary.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java [hunk 12]
```diff
@@ -283,6 +339,48 @@ public void alterTable(
...
+        CatalogTable table = getTable(tablePath);
+
+        validateAlterTable(table, (CatalogTable) newTable);
+
+        List<SchemaChange> changes = new ArrayList<>();
+        if (null != tableChanges) {
+            List<SchemaChange> schemaChanges =
+                    tableChanges.stream()
+                            .map(this::toSchemaChange)
+                            .filter(Objects::nonNull)
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/FlinkCatalog.java:362]: Seems `toSchemaChange` won't return null, so we don't need to filter.

File: paimon-flink/paimon-flink-1.16/src/test/java/org/apache/paimon/flink/CatalogITCaseBase.java [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.flink.streaming.api.environment.ExecutionCheckpointingOptions.CHECKPOINTING_INTERVAL;
+
+/** ITCase for catalog. */
+public abstract class CatalogITCaseBase extends AbstractTestBase {
```
Comment (yuzelin) [inline paimon-flink/paimon-flink-1.16/src/test/java/org/apache/paimon/flink/CatalogITCaseBase.java:56]: Please remove unused method:
`batchSql`, `streamSqlIter`, `getTableDirectory` and `findLatestSnapshot`.

**Review by yuzelin — APPROVED** [review]
LGTM. Only need some minor improve.

File: paimon-flink/paimon-flink-1.16/src/test/java/org/apache/paimon/flink/CatalogITCaseBase.java (hunk 13 above)
Comment (zhangjun0x01) [reply paimon-flink/paimon-flink-1.16/src/test/java/org/apache/paimon/flink/CatalogITCaseBase.java:48]: I update it

**Review by JingsongLi — APPROVED** [review]
+1

**Comment by JingsongLi** [comment]
wait for testing...
